### PR TITLE
fix(code_mode): normalize export-default arrow with a prologue

### PR DESCRIPTION
**Pull request by jmagar**
## Why

Follow-up to #90. Code Mode normalization mishandled an **arrow function in `export default` position when preceded by a prologue**, e.g.:

```js
const x = 1;
export default async () => x;
```

Two failure modes:
- **async arrow** → Boa's `parse_module` cannot parse an arrow in default-export position (`expected token 'function', got '('`), so both the module and script parses failed and the input **loose-wrapped into invalid JS** (left `export default` inside the wrapper body).
- **plain/expr/function defaults with a prologue** → Boa *does* parse these, but its AST arms (`DefaultAssignmentExpression` / function / class) render only the export declaration and **drop the prologue**, emitting valid-but-broken JS (the default's free variables are `undefined` at runtime).

This was disclosed as a known limitation on #90.

## What

Handle `{prologue} export default <value>` **textually, before the AST path**: split off the default-export value at a statement boundary, re-derive the no-prologue entry (reusing the existing start-anchored `export default` handling), then run the prologue first and invoke that entry inside one wrapper so it **closes over the prologue's bindings**:

```js
// const x = 1; export default async () => x   ⇒
async () => {
  const x = 1;
  return await (async () => x)();
}
```

Conservative: the split only fires when a real prologue ends at a `;`/`}` boundary; the start-anchored (no-prologue) case is unchanged.

## Tests

- `normalize_user_code_wraps_export_default_arrow_with_prologue` (async)
- `normalize_user_code_wraps_export_default_plain_arrow_with_prologue` (non-async)
- `normalized_export_default_arrow_with_prologue_executes_end_to_end` — **non-vacuous**: the arrow references a prologue `const`, so a dropped prologue would surface no `tool_call` and the round-trip helper would fail.

The existing 10a test (`normalize_user_code_finds_export_default_after_prologue`) still passes — and now actually preserves the prologue rather than emitting valid-but-broken JS.

Verified: 44 code_mode/normalize tests pass; clippy `-D warnings` + fmt clean.

---
## Summary by cubic
Fixes normalization for `export default` arrows and functions that follow a prologue. Emits valid JS, preserves prologue bindings (including named exports), drops imports inside the wrapper, and avoids regressions with trailing exports and comments.

- **Bug Fixes**
  - Async arrow default fallback: runs only after both module and script parses fail; splits at a real `;`/`}` boundary; tolerates trailing comments and `//` inside strings via a suffix-only strip.
  - AST path keeps the prologue: render non-export statements; keep binding-carrying named exports with `export` stripped; skip imports; select only the default export so later named exports don’t shadow it.
  - Textual fallback reparses the prologue to strip `export` from named exports and now drops `import` declarations too.
  - Anonymous default functions: drop Boa’s synthesized `default` name when IIFE-wrapping, anchored at the leading `function`.
  - Tests: URL+trailing-comment boundary, export var/function bindings, async-arrow + named-export e2e, import-only prologue e2e, and multi-statement prologue e2e.

<sup>Written for commit 909ee6141e0297919e0fc8e9271d6e93b3cf9be6. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/jmagar/lab/pull/92?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Improved normalization for scripts with a leading prologue followed by a default export: preserves prologue bindings and produces a single async wrapper that invokes the export, with a conservative textual fallback when parsing fails.

* **Refactor**
  * Reworked module normalization to reliably extract and wrap default-export entries while preserving leading bindings and avoiding synthesized-name issues.

* **Tests**
  * Added unit and end-to-end tests for prologue+default-export scenarios and edge cases (exports in strings, trailing comments).

### CODE REVIEW

**Comment by jmagar** [comment]
Addressed all 4 review threads in `8d732d5c`:

- **Parse before textual split** (chatgpt L61, cubic L80): the textual export-default split now runs only **after** both the module and script parses fail. Valid script code containing the literal `"; export default "` (e.g. in a string) parses normally first and is never mis-split. Added a guard test (`const s = "; export default "; s` → returns `s`).
- **Preserve prologue via AST** (the broader fix): `normalize_module_code` now renders the non-export statements (skipping imports) and invokes the export entry so it closes over them — fixing the latent prologue-drop for the DefaultAssignmentExpression / function / class arms too, without any string matching.
- **Comment-aware boundary** (cubic L83 / coderabbit L85, P1): the textual fallback's statement-boundary check skips a trailing line/block comment, so `const x = 1; // note\nexport default async () => x` splits correctly. Added a test.

46 normalize tests pass (incl. the non-vacuous e2e runner test); clippy `-D warnings` + fmt clean.

**Comment by jmagar** [comment]
Addressed the new review in `452bee24`:

**[P2] normalize.rs — last-export-wins (cubic):** fixed. The module-item scan kept the *last* `ExportDeclaration`, so a named export after `export default` (e.g. `const x = 1; export default x; export const y = 2`) shadowed the default and the entry match bailed to an invalid loose-wrap. Now captures the **default** export specifically (new `is_default_export` helper); named exports + imports are skipped. Added a regression test (prologue + default + trailing named export — the form that actually reaches the module scan; a no-prologue `export default …` takes the start-anchored strip path).

**`return n` thread (suggested `contains("() => n")`):** respectfully not applied — it's based on a misread of the code path. The plain-arrow default goes through the `DefaultAssignmentExpression` arm, which re-prints the arrow via `expr.to_interned_string()` to `() => { return n; }` (block body). So the output **does** contain `return n` and the test passes; the suggested `contains("() => n")` would *fail* because the verbatim `() => n` form isn't what's emitted. Confirmed by the passing unit test and the runtime e2e (`normalized_export_default_plain_arrow_with_prologue_executes_end_to_end`).

51 normalize/runner tests pass; clippy `-D warnings` + fmt clean.

File: crates/lab/src/dispatch/gateway/code_mode/normalize.rs [hunk 1]
```diff
@@ -45,7 +51,62 @@ pub fn normalize_user_code(code: &str) -> String {
...
+        let before = code[..idx].trim_end();
+        // A real statement terminator wins as-is. Only when `before` does not
+        // already end at a boundary do we retry after stripping a trailing
+        // comment, so `const x = 1; // note\nexport default ...` still splits —
+        // without letting a `//` *inside* a prologue string (e.g. a "http://"
+        // URL) corrupt an otherwise-terminated prologue.
+        let ends_at_boundary = |s: &str| s.ends_with(';') || s.ends_with('}');
+        let comment_stripped = strip_trailing_comment(before).trim_end();
+        if (!before.is_empty() && ends_at_boundary(before))
+            || (!comment_stripped.is_empty() && ends_at_boundary(comment_stripped))
```
[the remark this answers is not in the document]
Comment (jmagar) [reply crates/lab/src/dispatch/gateway/code_mode/normalize.rs:115]: Fixed in dc91b2e5. Added `strip_suffix_line_comment`, used inline in `split_prologue_export_default`: it inspects only the last `//` on the final line, so a prologue with both a `"http://..."` URL string and a real trailing comment now strips the genuine comment and splits at the `;` boundary. The shared `strip_trailing_comment` (and its other caller) is untouched. The downstream `;`/`}` boundary check plus the both-parses-failed precondition gate any spurious split. Regression test: `normalize_user_code_export_default_arrow_url_string_and_trailing_comment_prologue`. Note: the `*/` block-comment branch shares the same `count==1` conservatism but is left as-is — a block comment in a prologue tail before an async-arrow default is more exotic still.

File: crates/lab/src/dispatch/gateway/code_mode/normalize.rs [hunk 2]
```diff
@@ -86,44 +144,89 @@ fn normalize_user_code_parsed(source: &str) -> Option<String> {
...
+            // re-export) after `export default` must not shadow the default and
+            // cause normalization to bail. Named exports have no entry/runtime
+            // role in the sandbox, so they (and imports) are skipped.
+            boa_ast::ModuleItem::ExportDeclaration(decl) if is_default_export(decl) => {
+                export = Some(decl);
+            }
+            boa_ast::ModuleItem::StatementListItem(stmt) => {
+                prologue.push(stmt.to_indented_string(&interner, 0));
+            }
+            boa_ast::ModuleItem::ExportDeclaration(_)
```
[the remark this answers is not in the document]
Comment (jmagar) [reply crates/lab/src/dispatch/gateway/code_mode/normalize.rs]: Fixed in dc91b2e5. Binding-carrying named exports (`export const/let/var/function/class`) now render into the prologue with the `export` keyword stripped, via the new shared `render_prologue_item`, so a default that references one keeps the binding at runtime. `export {}` lists, re-exports, and imports still drop (no sandbox-runtime role). The same gap existed on the textual async-arrow fallback (boa can't parse an async-arrow `export default`, so the prologue was recovered textually with `export` intact — a syntax error in the wrapper); `strip_prologue_exports` reparses the prologue alone and re-renders without `export`. Tests: module-path binding kept, `export var`/`export function` bindings, plus an end-to-end run of `export const tool = ...; export default async () => callTool(tool, {})`.

File: crates/lab/src/dispatch/gateway/code_mode/normalize.rs [hunk 3]
```diff
@@ -86,47 +160,167 @@ fn normalize_user_code_parsed(source: &str) -> Option<String> {
...
+        return prologue.to_string();
+    };
+    // Only rewrite when the prologue actually carries a named export; otherwise
+    // keep the verbatim text so re-rendering never perturbs the common case.
+    let has_named_export = module
+        .items()
+        .items()
+        .iter()
+        .any(|item| matches!(item, boa_ast::ModuleItem::ExportDeclaration(_)));
+    if !has_named_export {
```
[the remark this answers is not in the document]
Comment (jmagar) [reply crates/lab/src/dispatch/gateway/code_mode/normalize.rs]: Fixed in 909ee614. Good catch — regression in the new `strip_prologue_exports`. The guard checked for a *named export*, so an import-only prologue (`import x from "y"; export default async () => ...`) was returned verbatim, leaving the `import` inside the wrapper. Widened the guard to "any non-statement item": named exports get `export` stripped, imports are dropped (no module loader in the sandbox — same as `normalize_module_code` already does). Pure-statement prologues stay verbatim. Regression test: `normalize_user_code_drops_import_in_async_arrow_default_prologue`.
